gix-fs/tests/stack/mod.rs

```diff
@@ -26,6 +27,287 @@ impl gix_fs::stack::Delegate for Record {
+fn path_join_handling() {
+    let absolute = p("/absolute");
+    assert!(
+        absolute.is_relative(),
+        "on Windows, absolute Linux paths are considered relative (and relative to the current drive)"
+    );
+    let bs_absolute = p("\\absolute");
+    assert!(
+        absolute.is_relative(),
+        "on Windows, strange single-backslash paths are relative (and relative to the current drive)"
+    );
+    assert_eq!(
+        p("relative").join(absolute),
+        absolute,
+        "relative + absolute = absolute - however, they kind of act like they are absolute in conjunction with relative base paths"
+    );
+    assert_eq!(
+        p("relative").join(bs_absolute),
+        bs_absolute,
+        "relative + absolute = absolute - backslashes aren't special here, and it just acts like it's absolute"
+    );
+
+    assert_eq!(
+        p("c:").join("relative"),
+        p("c:relative"),
+        "drive + relative = strange joined result with missing backslash, but it's a valid path that works just like `c:\relative`"
+    );
+    assert_eq!(
+        p("c:\\").join("relative"),
+        p("c:\\relative"),
+        "absolute + relative = joined result"
+    );
+
+    assert_eq!(
+        p("\\\\?\\base").join(absolute),
+        p("\\\\?\\base\\absolute"),
+        "absolute1 + unix-absolute2 = joined result with backslash"
+    );
+    assert_eq!(
+        p("\\\\.\\base").join(absolute),
+        p("\\\\.\\base\\absolute"),
+        "absolute1 + absolute2 = joined result with backslash (device namespace)"
+    );
+    assert_eq!(
+        p("\\\\?\\base").join(bs_absolute),
+        p("\\\\?\\base\\absolute"),
+        "absolute1 + absolute2 = joined result"
+    );
+    assert_eq!(
+        p("\\\\.\\base").join(bs_absolute),
+        p("\\\\.\\base\\absolute"),
+        "absolute1 + absolute2 = joined result (device namespace)"
+    );
+
+    assert_eq!(p("/").join("C:"), p("C:"), "unix-absolute + win-drive = win-drive");
+    assert_eq!(
+        p("d:/").join("C:"),
+        p("C:"),
+        "d-drive + c-drive = c-drive - interesting, as C: is supposed to be relative"
+    );
+    assert_eq!(
+        p("d:\\").join("C:\\"),
+        p("C:\\"),
+        "d-drive-with-bs + c-drive-with-bs = c-drive-with-bs - nothing special happens with backslashes"
+    );
+    assert_eq!(
+        p("c:\\").join("\\\\.\\"),
+        p("\\\\.\\"),
+        "c-drive-with-bs + device-namespace-unc = device-namespace-unc"
+    );
+    assert_eq!(
+        p("/").join("C:/"),
+        p("C:\\"),
+        "unix-absolute + win-drive = win-drive, strangely enough it changed the trailing slash to backslash, so better not have trailing slashes"
+    );
+    assert_eq!(p("/").join("C:\\"), p("C:\\"), "unix-absolute + win-drive = win-drive");
+    assert_eq!(
+        p("\\\\.").join("C:"),
+        p("C:"),
+        "device-namespace-unc + win-drive-relative = win-drive-relative - c: was supposed to be relative, but it's not acting like it."
```

Comment:
The `device-namespace-unc` part of the custom message suggests that this may have been meant to start with `p("\\\\.\\")` rather than `p("\\\\.")`. As currently written, I am not sure what the ideal behavior would be here, because I am not sure the trailing `\` of the prefix `\\?\` is strictly speaking functioning as a path separator. The Rust standard library [does sometimes treat `\\.\` and `\\.` differently](https://github.com/EliahKagan/pathjoin-unc) as the left side of a path join. However, both for the relative path `C:` and the absolute path `C:\` on the right, the result is the same.

However, for the same reason as in the above comment, that `C:` is relative should not be a reason to expect anything other than this behavior. It's less clear what it means to be *located at* `\\.` or `\\.\`. But I don't think it should be an exception to the usual effect of `C:` on the right side.

Furthermore, although `\\.\C:` and `\\.\C:\` are both valid paths, we *cannot* correctly form either one of them by concatenating `\\.` or `\\.\` with `C:`, and the reason is *because* a path `C:` is a relative path:

- `\\.\C:` is in the device namespace and ends in `C:` with no path separator, so that `C:` is a *device name*. Therefore `\\.\C:` refers to the `C:` drive as a device; the `C:` in `\\.\C:` is not a reference to the current directory, root directory, or *any* file or directory on the `C:` drive, but instead to the `C:` *drive* itself. (`\\.\C:` is an absolute path, but not to a file or directory.)
- `\\.\C:\`, due to the trailing `\`, is the root directory on the `C:` drive, and thus generally equivalent to `\\?\C:\` (though I don't know that all APIs that accept one will accept the other or that all APIs that accept both will necessarily treat them identically). But this is an absolute path to the root of the `C:` drive, which `C:` by itself does not designate because it is a relative path whose meaning depends on the current directory on the `C:` drive.